File: csharp/BCLifeHash/BCLifeHash/BitEnumerator.cs
namespace BlockchainCommons.BCLifeHash;

internal sealed class BitEnumerator
{
    private readonly byte[] _data;
    private int _index;
    private byte _mask;

    public BitEnumerator(byte[] data)
    {
        _data = data;
        _index = 0;
        _mask = 0x80;
    }

    public bool HasNext => _mask != 0 || _index != _data.Length - 1;

    public bool Next()
    {
        if (!HasNext)
            throw new InvalidOperationException("BitEnumerator underflow");

        if (_mask == 0)
        {
            _mask = 0x80;
            _index++;
        }

        var b = (_data[_index] & _mask) != 0;
        _mask >>= 1;
        return b;
    }

    public uint NextUint2()
    {
        uint bitMask = 0x02;
        uint value = 0;
        for (var i = 0; i < 2; i++)
        {
            if (Next())
                value |= bitMask;
            bitMask >>= 1;
        }
        return value;
    }

    public uint NextUint8()
    {
        uint bitMask = 0x80;
        uint value = 0;
        for (var i = 0; i < 8; i++)
        {
            if (Next())
                value |= bitMask;
            bitMask >>= 1;
        }
        return value;
    }

    public uint NextUint16()
    {
        uint bitMask = 0x8000;
        uint value = 0;
        for (var i = 0; i < 16; i++)
        {
            if (Next())
                value |= bitMask;
            bitMask >>= 1;
        }
        return value;
    }

    public double NextFrac()
    {
        return NextUint16() / 65535.0;
    }

    public void ForAll(Action<bool> f)
    {
        while (HasNext)
            f(Next());
    }
}

internal sealed class BitAggregator
{
    private readonly List<byte> _data = new();
    private byte _bitMask;

    public void Append(bool bit)
    {
        if (_bitMask == 0)
        {
            _bitMask = 0x80;
            _data.Add(0);
        }

        if (bit)
            _data[^1] |= _bitMask;

        _bitMask >>= 1;
    }

    public byte[] GetData()
    {
        return _data.ToArray();
    }
}

File: kotlin/bc-lifehash/src/main/kotlin/com/blockchaincommons/bclifehash/BitEnumerator.kt
package com.blockchaincommons.bclifehash

class BitEnumerator(
    private val data: ByteArray,
) {
    private var index: Int = 0
    private var mask: Int = 0x80

    fun hasNext(): Boolean = mask != 0 || index != data.size - 1

    fun next(): Boolean {
        require(hasNext()) { "BitEnumerator underflow" }

        if (mask == 0) {
            mask = 0x80
            index += 1
        }

        val bit = (data[index].toInt() and mask) != 0
        mask = mask ushr 1
        return bit
    }

    fun nextUInt2(): Int {
        var bitMask = 0x02
        var value = 0
        repeat(2) {
            if (next()) {
                value = value or bitMask
            }
            bitMask = bitMask ushr 1
        }
        return value
    }

    fun nextUInt8(): Int {
        var bitMask = 0x80
        var value = 0
        repeat(8) {
            if (next()) {
                value = value or bitMask
            }
            bitMask = bitMask ushr 1
        }
        return value
    }

    fun nextUInt16(): Int {
        var bitMask = 0x8000
        var value = 0
        repeat(16) {
            if (next()) {
                value = value or bitMask
            }
            bitMask = bitMask ushr 1
        }
        return value
    }

    fun nextFrac(): Double = nextUInt16().toDouble() / 65535.0

    fun forAll(block: (Boolean) -> Unit) {
        while (hasNext()) {
            block(next())
        }
    }
}

class BitAggregator {
    private val data: MutableList<Byte> = mutableListOf()
    private var bitMask: Int = 0

    fun append(bit: Boolean) {
        if (bitMask == 0) {
            bitMask = 0x80
            data.add(0)
        }

        if (bit) {
            val last = data.lastIndex
            data[last] = (data[last].toInt() or bitMask).toByte()
        }

        bitMask = bitMask ushr 1
    }

    fun data(): ByteArray = data.toByteArray()
}
